### PR TITLE
Enable CPU FlexAttention on AArch64

### PR DESCRIPTION
**Pull request by Anndrey24**
## Summary

This PR extends CPU FlexAttention support to AArch64 across supported Arm CPU capabilities, including SVE128 and SVE256.

It also optimizes the fused softmax path, mainly improving BF16 performance.

## Testing

On an Arm Neoverse-V2 SVE128 system, the BF16 softmax optimization shows the following improvements over the baseline:

| Pattern          | Sequence length | Improvement |
| ---------------- | --------------: | ----------: |
| Dense            |           2,048 |        9.8% |
| Dense            |           8,192 |        4.1% |
| Causal           |           2,048 |       12.2% |
| Causal           |           8,192 |       12.0% |
| Local window 128 |           2,048 |       17.8% |
| Local window 128 |           8,192 |       17.7% |

For sparse local-window attention at sequence length 8,192, FlexAttention also shows the following speedups compared with FlashAttention:

| dtype | FlexAttention speedup |
| ----- | --------------------: |
| BF16  |                13.09x |
| FP32  |                23.00x |

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @aditew01 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Review by fadara01 — COMMENTED** [review]
Nice!

Why does this need to be specific for SVE256?

File: torch/_inductor/codegen/cpp_flex_attention_template.py [hunk 1]
```diff
@@ -312,6 +312,7 @@
 
 FLEX_ATTENTION_TEMPLATE = r"""
   bool need_pack = false;
+#if !defined(CPU_CAPABILITY_SVE256)
```
Comment (fadara01) [inline torch/_inductor/codegen/cpp_flex_attention_template.py]: `is_reduced_type` is constexpr, why don't we use it to decide on pack+brgemm vs blas_gemm like we do in flash attention? See: https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/cpu/FlashAttentionKernel.cpp#L599C25-L599C63

**Comment by Anndrey24** [comment]
> Nice!
> 
> Why does this need to be specific for SVE256?

Thanks! I scoped this PR to SVE256 for now because `VecSVE256` is the existing SVE ISA support in Inductor at the moment ([link](https://github.com/pytorch/pytorch/blob/b671f8ab97a88399d1a6bdbc32cfdeeaadba1932/torch/_inductor/cpu_vec_isa.py#L173)).

I believe there is already work to extend `cpu_vec_isa.py` for a more general SVE in #184555, and adding SVE128 support to FlexAttention afterwards should be relatively straightforward.

**Comment by aditew01** [comment]
@claude review the PR

File: torch/_inductor/kernel/flex/flex_cpu.py [hunk 2]
```diff
@@ -32,12 +32,12 @@ def check_cpu_supported():
     requires_avx2_on_cpu = (
         torch.cpu._is_avx2_supported() and os.getenv("ATEN_CPU_CAPABILITY") != "default"
     )
-    requires_sve256_on_cpu = (
-        torch.backends.cpu.get_cpu_capability() == "SVE256"
+    requires_sve_on_cpu = (
```
Comment (fadara01) [inline torch/_inductor/kernel/flex/flex_cpu.py]: any reason why this requires SVE and doesn't work with Neon for example?

File: torch/_inductor/codegen/cutedsl/cutedsl_kernel.py [hunk 3]
```diff
@@ -120,7 +120,7 @@ class CuteDSLSubgraphInfo:
     body: IndentedBuffer
     template_mask: str | None = None
     template_out: str | None = None
-    cse: CSE[Any] | None = None
+    cse: CSE[Any, Any] | None = None
```
Comment (fadara01) [inline torch/_inductor/codegen/cutedsl/cutedsl_kernel.py]: why is this change needed?

**Comment by fadara01** [comment]
@claude review the PR

**Comment by zxiiro** [comment]
I think this PR needs to be rebased. Its jobs are asking for runners that don't exist. Can you rebase it and see if it picks up valid runners. The runners should be prefixed with `lf-` rather than `lf.`.

File: torch/testing/_internal/common_device_type.py [hunk 4]
```diff
@@ -2371,10 +2371,14 @@ def get_all_device_types() -> list[str]:
...
 
-# skip since currently flex attention requires at least `avx2` support on CPU.
+# skip since currently flex attention requires at least `avx2`, SVE, or NEON support on CPU.
 IS_FLEX_ATTENTION_CPU_PLATFORM_SUPPORTED = (
     not IS_MACOS
-    and torch.cpu._is_avx2_supported()
+    and (
+        torch.cpu._is_avx2_supported()
+        or torch.cpu._is_sve_supported()
+        or torch.cpu._is_neon_supported()
```
Comment (fadara01) [inline torch/testing/_internal/common_device_type.py]: isn't something like `is_aarch64()` enough to indicate that it's supported for all Arm platforms.

**Comment by aditew01** [comment]
@claude review the PR and comment if the failures are related

File: torch/cpu/__init__.py [hunk 5]
```diff
@@ -83,6 +83,21 @@ def _is_avx2_supported() -> bool:
...
 
 
+def _is_sve_supported() -> bool:
+    r"""Returns a bool indicating if CPU supports SVE."""
+    return torch.backends.cpu.get_cpu_capability() in ("SVE128", "SVE256")
+
+
+def _is_aarch64_supported() -> bool:
+    r"""Returns a bool indicating if CPU is AArch64."""
+    return get_capabilities().get("architecture") in ("arm64", "aarch64")
```
Comment (aditew01) [inline torch/cpu/__init__.py]: add "ARM64" 
ref: https://github.com/pytorch/pytorch/pull/189346/changes

File: torch/_inductor/codegen/cutedsl/cutedsl_kernel.py (hunk 3 above)
Comment (Anallear) [reply torch/_inductor/codegen/cutedsl/cutedsl_kernel.py]: changed back to  CSE[Any]

File: torch/_inductor/kernel/flex/flex_cpu.py (hunk 2 above)
Comment (Anallear) [reply torch/_inductor/kernel/flex/flex_cpu.py]: now it working with neon, sve128, sve256

File: torch/testing/_internal/common_device_type.py (hunk 4 above)
Comment (Anallear) [reply torch/testing/_internal/common_device_type.py]: modified to be more general

**Review by aditew01 — APPROVED** [review]

File: torch/_inductor/kernel/flex/flex_cpu.py [hunk 6]
```diff
@@ -32,8 +32,12 @@ def check_cpu_supported():
     requires_avx2_on_cpu = (
         torch.cpu._is_avx2_supported() and os.getenv("ATEN_CPU_CAPABILITY") != "default"
     )
+    requires_arm_on_cpu = (
+        torch.cpu._is_aarch64_supported()
+        and os.getenv("ATEN_CPU_CAPABILITY") != "default"
```
Comment (fadara01) [inline torch/_inductor/kernel/flex/flex_cpu.py:37]: I think default capability means NEON, why is this not active for NEON?

**Review by fadara01 — APPROVED** [review]

**Comment by fadara01** [comment]
@drisspg @jgong5 @leslie-fang-intel @Skylion007 could you please take a look?

**Comment by fadara01** [comment]
@claude review
